### ISSUE TITLE
A few questions about the training pipeline

### ISSUE THREAD

**Issue by avipartho** [kohjingyu/gill#5]
Hi,

I read your paper and it was a great work! Thanks for sharing your codebase with the community. As I was going through your codes, I came across a few places, where I would greatly appreciate your explanations/suggestions. Here are my questions - 

- What does the CE loss from [here](https://github.com/kohjingyu/gill/blob/51f3d98a2b3a0d9776e221cf3c3b5331326c0dc4/main.py#L506) stand for, i.e. which of the 4 losses from the paper it refers to?
- Does the CE loss from [here](https://github.com/kohjingyu/gill/blob/51f3d98a2b3a0d9776e221cf3c3b5331326c0dc4/main.py#L508) refer to the l_p loss in the paper? If not, which loss form the paper it refers to?
- From these lines ([line1](https://github.com/kohjingyu/gill/blob/c1d33098fba2e5dc30f3f940449af1469ce549ab/gill/models.py#L285), [line2](https://github.com/kohjingyu/gill/blob/c1d33098fba2e5dc30f3f940449af1469ce549ab/gill/models.py#L293), [line3](https://github.com/kohjingyu/gill/blob/c1d33098fba2e5dc30f3f940449af1469ce549ab/gill/models.py#L361)), it looks like all tokens that are not part of the caption text or `[IMG0]` have been set to -100 to be ignored from calculating loss. Is my understanding correct? If it is, how are we learning embeddings for other `[IMG{r}]` tokens (`r={2,3,...,8}`)?

**Comment by kohjingyu** [comment]
Thanks for your kind words!

> What does the CE loss from [here](https://github.com/kohjingyu/gill/blob/51f3d98a2b3a0d9776e221cf3c3b5331326c0dc4/main.py#L506) stand for, i.e. which of the 4 losses from the paper it refers to?

> Does the CE loss from [here](https://github.com/kohjingyu/gill/blob/51f3d98a2b3a0d9776e221cf3c3b5331326c0dc4/main.py#L508) refer to the l_p loss in the paper? If not, which loss form the paper it refers to?

This and the loss defined on line 508 make up $l_p$ in the paper (equation 2). This is the loss for training the model to produce the [IMG] tokens at the end of "caption-like" text. Both L506 and L508 are actually the same loss, since the same caption (e.g., "A picture of a dog [IMG0]...[IMG7]") are used for retrieval and generation. This is why they have a 0.5 multiplier, so that they sum to be $l_p$.


> From these lines ([line1](https://github.com/kohjingyu/gill/blob/c1d33098fba2e5dc30f3f940449af1469ce549ab/gill/models.py#L285), [line2](https://github.com/kohjingyu/gill/blob/c1d33098fba2e5dc30f3f940449af1469ce549ab/gill/models.py#L293), [line3](https://github.com/kohjingyu/gill/blob/c1d33098fba2e5dc30f3f940449af1469ce549ab/gill/models.py#L361)), it looks like all tokens that are not part of the caption text or [IMG0] have been set to -100 to be ignored from calculating loss. Is my understanding correct? If it is, how are we learning embeddings for other [IMG{r}] tokens (r={2,3,...,8})?

That's right, and the reason for this is that we [force the generation](https://github.com/kohjingyu/gill/blob/c1d33098fba2e5dc30f3f940449af1469ce549ab/gill/models.py#L517-L520) of the `r={2,3,...,8}` tokens as the next 7 tokens whenever the model produces `[IMG0]` (since we _always_ need all 8 tokens for generation/retrieval, so it doesn't make sense to have a partial set of the `[IMG]` tokens). The embeddings of `[IMG2]...[IMG8]` tokens are therefore only learnt through the other losses (in particular the generation loss $l_g$), when their embeddings/hidden states are used for computing the generation/retrieval objectives. $l_p$ doesn't affect `[IMG2]...[IMG8]` tokens. So the model will never produce `[IMG2]...[IMG8]` organically, but their representations are still helpful for feeding into the GILLMapper module for image generation.


Hope that makes sense!

**Comment by avipartho** [comment]
Thanks for your quick response. Reopening this issue for another query regarding the pipeline (didn't want to unnecessarily create new issue).

If I am not wrong, [this line](https://github.com/kohjingyu/gill/blob/c1d33098fba2e5dc30f3f940449af1469ce549ab/gill/models.py#L75) makes the entire OPT embedding layer trainable. It is also evident from the `param_count.txt` file your scripts generate. However, according to the paper only the `[IMG]` embedding matrix **E<sub>img</sub>** was supposed to be trainable. Did I miss anything here?

First few lines from param_count.txt : 

Module                                                       | Trainable  |           Shape |  Param Count |
----------------------------------------------------------------------------------------------------------------------------------
| model.logit_scale                                            | True       |              () |            1 |
| model.lm.model.decoder.embed_tokens.weight                   | True       |   (50274, 4096) |  205,922,304 |

**Comment by avipartho** [comment]
Thanks for your quick response. Reopening this issue for another query regarding the pipeline (didn't want to unnecessarily create new issue).

If I am not wrong, [this line](https://github.com/kohjingyu/gill/blob/c1d33098fba2e5dc30f3f940449af1469ce549ab/gill/models.py#L75) makes the entire OPT embedding layer trainable. It is also evident from the `param_count.txt` file your scripts generate. However, according to the paper only the `[IMG]` embedding matrix **E<sub>img</sub>** was supposed to be trainable. Did I miss anything here?

First few lines from param_count.txt : 

Module                                                       | Trainable  |           Shape |  Param Count |
----------------------------------------------------------------------------------------------------------------------------------
| model.logit_scale                                            | True       |              () |            1 |
| model.lm.model.decoder.embed_tokens.weight                   | True       |   (50274, 4096) |  205,922,304 |

**Comment by kohjingyu** [comment]
You're right, they become trainable, which is why we zero out the gradients of the non-[IMG] embeddings here:

https://github.com/kohjingyu/gill/blob/53fdcf22952ba0e08fe2cf5b006948df7c5636aa/main.py#L578-L587

This is not super ideal, but I think it is overall cleaner than concatenating a trainable embedding matrix with a frozen one.

**Comment by avipartho** [comment]
Thanks again. Unfortunately, I missed this section of the script. 

Is it also correct to say that for the _l<sub>p</sub>_ loss, you are considering the loss for generating each token of the input text (caption) i.e. the negative log likelihood of generating token _s<sub>t</sub>_ conditioned on _s<sub>1</sub>,...,s<sub>t-1</sub>_ where t={1,...,T}?

**Comment by kohjingyu** [comment]
Yes that's right!

**Comment by avipartho** [comment]
In that case, I believe equation 2 is slightly misleading, as the summation goes over `i` from `1` to `r` there. This practically says that we are considering loss for generating all 8 `[IMG]` tokens.

**Comment by avipartho** [comment]
As I was trying to train my own model and run inference using the saved checkpoint, I noticed a few things, please verify (might be helpful for other users).

- Because `GILLArgs()` only has local variables and no attributes, all of them don't get [saved](https://github.com/kohjingyu/gill/blob/51f3d98a2b3a0d9776e221cf3c3b5331326c0dc4/main.py#L277) in the `model_args.json` file unless specifically set after instantiating. One such attribute is `text_emb_layers`. Turning all local variables into class attributes can solve this.
- Currently the `main.py` script also [saves the scheduler state](https://github.com/kohjingyu/gill/blob/51f3d98a2b3a0d9776e221cf3c3b5331326c0dc4/main.py#L419), which pretty much saves the entire model ([probable reason](https://discuss.pytorch.org/t/scheduler-state-dict-size/100928/2)) and therefore results in a large checkpoint. 
- The pretrained checkpoint provided in this codebase has a shape of (8,4096) for `input_embeddings.weight`  whereas running the `main.py` will produce a checkpoint with `input_embeddings.weight` of shape (50274, 4096). Looks like the provided checkpoint contains only the trainable [IMG] token embeddings. This requires either changing [this line](https://github.com/kohjingyu/gill/blob/c1d33098fba2e5dc30f3f940449af1469ce549ab/gill/models.py#L889) or [this line](https://github.com/kohjingyu/gill/blob/c1d33098fba2e5dc30f3f940449af1469ce549ab/gill/models.py#L881) to run inference with the produced checkpoint. For example, 
```python
img_token_embeddings = state_dict['model.input_embeddings.weight'].cpu().detach()[-model_kwargs['num_tokens']:, :]
```

**Comment by kohjingyu** [comment]
Thanks for sharing this!

> * The pretrained checkpoint provided in this codebase has a shape of (8,4096) for `input_embeddings.weight`  whereas running the `main.py` will produce a checkpoint with `input_embeddings.weight` of shape (50274, 4096). Looks like the provided checkpoint contains only the trainable [IMG] token embeddings. This requires either changing [this line](https://github.com/kohjingyu/gill/blob/c1d33098fba2e5dc30f3f940449af1469ce549ab/gill/models.py#L889) or [this line](https://github.com/kohjingyu/gill/blob/c1d33098fba2e5dc30f3f940449af1469ce549ab/gill/models.py#L881) to run inference with the produced checkpoint. For example,

You're right, and I also realized that I hadn't uploaded the script used to prune the checkpoints (keeping just the  trained weights, and discarding the pretrained model weights). I just did that here: https://github.com/kohjingyu/gill/blob/main/scripts/prune_model_ckpt.py

I think this is essentially the same as the changes you probably made locally, though I haven't tested this script in a while.

**Comment by avipartho** [comment]
Thanks for sharing the script! Just noticed a few things - 

- [These arguments](https://github.com/kohjingyu/gill/blob/4dd1c22ac9bec7d4c0e2a04c56f0e10bf405fe87/scripts/prune_model_ckpt.py#L23) no longer exist in the current version. Could it be that you probably coalesced them into `num_tokens`? Please verify.
- This [line](https://github.com/kohjingyu/gill/blob/4dd1c22ac9bec7d4c0e2a04c56f0e10bf405fe87/scripts/prune_model_ckpt.py#L32) gives error, as it is trying to mutate an ordered dict during the for loop. This can be avoided by making an empty dict first and then copying everything there (just like it's done [here](https://github.com/kohjingyu/gill/blob/7653cf5ce2df5297c53e94958420bb7814a3870a/gill/models.py#L877C3-L880C47) in `models.py`)
- I believe [the example usage](https://github.com/kohjingyu/gill/blob/4dd1c22ac9bec7d4c0e2a04c56f0e10bf405fe87/scripts/prune_model_ckpt.py#L4) should be `python scripts/prune_model_ckpt.py  runs/gill_exp`, given the location of the script.
- What's the use of `share_ret_gen`? I could not find any use of this in the `models.py`, `validate.py` or `main.py` script.

**Comment by kohjingyu** [comment]
Thanks for the notes! Sorry about this, it's what happens when you don't test before you upload.

> [These arguments](https://github.com/kohjingyu/gill/blob/4dd1c22ac9bec7d4c0e2a04c56f0e10bf405fe87/scripts/prune_model_ckpt.py#L23) no longer exist in the current version. Could it be that you probably coalesced them into num_tokens? Please verify.

That's right.

> What's the use of share_ret_gen? I could not find any use of this in the models.py, validate.py or main.py script.

`share_ret_gen` doesn't exist anymore, I think it was something used during debugging previously. I've updated the script as such, hopefully it works as expected now. Thanks for your help in debugging this!

**Comment by avipartho** [comment]
Another small update. I could not find `warmup-scheduler==0.3.2` (as mentioned in the requirements.txt file), the current available version is probably [0.3](https://pypi.org/project/warmup-scheduler/). Will it be compatible with your scripts? (I can verify that the training continues with this version)

**Comment by kohjingyu** [comment]
Ah, looks like it should be `pip install git+https://github.com/ildoonet/pytorch-gradual-warmup-lr.git` instead. The link you provided should still work though.

**Comment by avipartho** [comment]
I have another question. As mentioned above, the _l<sub>p</sub>_ loss includes the negative log likelihood (NLL) of generating each token of the input text (caption). Did you find this helpful for the overall model performance? I am asking this because from the name and purpose of this loss, I would assume that it was intended to only consider the NLL of generating [IMG] tokens.

**Comment by kohjingyu** [comment]
I have not run this particular ablation, sorry. I would guess that it does not have a significant effect on performance on the tasks we evaluated on.
